### PR TITLE
Added Google Analytics for ZLI

### PR DESCRIPTION
**Pull request by SaiprathapReddy7**
## Description of the change

Created Google Analytics account for ZLI to track the users, events and etc.. 

## Relevant release note information

Release Notes: Added Google Analytics for ZLI

## Related JIRA tickets

Relates to JIRA: CWC-1632

## Have you considered the security impacts?

Does this PR have any security impact?

- [ ] Yes
- [ ] No

If yes, please explain:

### CODE REVIEW

File: src/services/config/config.service.ts [hunk 1]
```diff
@@ -322,6 +333,10 @@ export class ConfigService implements ConfigInterface {
         }
     }
 
+    private async getGAToken(): Promise<string> {
+        return (await this.tokenHttpService.getGAToken()).token;
```
Comment (tflpd) [inline src/services/config/config.service.ts:337]: Maybe
```suggestion
        return (await this.tokenHttpService.getGAToken())?.token;
```
would be more safe

File: src/services/logger/logger.service.ts [hunk 2]
```diff
@@ -29,18 +30,31 @@ const loggingDebugFormat = printf(info => {
...
-        this.debugFlag = debug;
-        this.config = config;
-        this.silentFlag = silent;
+    private gaService: GAService
 
+    constructor(private config: LoggerConfigService, private debugFlag: boolean, private silentFlag: boolean, isStdEnabled: boolean) {
         // Build our logger
         this.buildLogger(isStdEnabled);
+
+        // Since the GA service is dependent on the logger, this will be set in the GA middleware in cil-driver
```
Comment (tflpd) [inline src/services/logger/logger.service.ts:40]: minor typo - 'cli-driver'*

File: src/services/v1/token/token.service.ts [hunk 3]
```diff
@@ -2,7 +2,7 @@ import { IdentityProvider } from '../../../../webshell-common-ts/auth-service/au
 import { ConfigService } from '../../config/config.service';
```
Comment (tflpd) [inline src/services/v1/token/token.service.ts:2]: No changes should be made in this file and no other component should use it as this is deprecated and will be deleted soon

File: src/services/v1/token/token.messages.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-export interface MixpanelTokenResponse
+export interface GATokenResponse
```
Comment (tflpd) [inline src/services/v1/token/token.messages.ts:1]: No changes should be made in this file and no other component should use it as this is deprecated and will be deleted soon

File: src/services/Tracking/Tracking.service.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+import { Logger } from '../logger/logger.service';
+const ua = require('universal-analytics');
+
+
+import { TargetType } from '../../../webshell-common-ts/http/v2/target/types/target.types';
+import { Dictionary } from 'lodash';
+import mixpanel, { Mixpanel } from 'mixpanel';
+import { TrackNewConnection } from './mixpanel.service.types';
+
+export class GAService
```
Comment (tflpd) [inline src/services/Tracking/Tracking.service.ts:11]: Please name the class the same as the file - or the file the same as the class; for filenames we use lower case letters

**Review by tflpd — CHANGES_REQUESTED** [review]

File: src/cli-driver.ts [hunk 6]
```diff
@@ -223,8 +226,41 @@ export class CliDriver
...
+            .middleware(async (argv) => {
+                if(!this.GACommands.has(baseCmd))
+                    return;
+                if(! this.configService.mixpanelToken()) {
+                    await this.configService.fetchMixpanelToken();
+                }
+                this.mixpanelService = mixpanelTrackingMiddleware(this.configService, argv);
+            })
+
+
```
[the remark this answers is not in the document]
Comment (SaiprathapReddy7) [reply src/cli-driver.ts]: Removed extra space

File: src/services/logger/logger.service.ts (hunk 2 above)
Comment (SaiprathapReddy7) [reply src/services/logger/logger.service.ts]: Updated typo error

File: src/handlers/middleware.handler.ts [hunk 7]
```diff
@@ -83,6 +84,19 @@ export function fetchDataMiddleware(configService: ConfigService, logger: Logger
     };
 }
 
+/*
+ * Helper function to get our our GA tracking middleware and track our cli command
```
[the remark this answers is not in the document]
Comment (SaiprathapReddy7) [reply src/handlers/middleware.handler.ts]: updated typo in the comment

File: src/services/Tracking/Tracking.service.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+{
+    private userId: string;
+    private visitor: any;
```
[the remark this answers is not in the document]
Comment (sidpremkumar) [reply src/services/Tracking/Tracking.service.ts:14]: We only expect one type: https://github.com/peaksandpies/universal-analytics

But I was unsure of how to extract the type from that module
2022-03-24 at 9 34 AM

Tried a couple of different things but no luck :/

File: src/services/Tracking/Tracking.service.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * Helper function to track a cli command.
+     * @param {string[]} args Args to the command
+    */
+    public async TrackCliCommand(args: string[]) {
+        const zliCommandCall = new Promise<void>(async (resolve, _) => {
+            await this.visitor.event('zli-command', this.baseCommand, (err: any) => {
+                if (err) {
+                    this.logger.error(`Error sending GA event zli-command: ${err}`);
```
[the remark this answers is not in the document]
Comment (sidpremkumar) [reply src/services/Tracking/Tracking.service.ts:46]: +1 moved to debug

File: src/services/Tracking/Tracking.service.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+                } else {
+                    this.logger.debug('Successfully tracked event');
+                }
+                resolve();
+            });
+        });
+        await zliCommandCall;
```
[the remark this answers is not in the document]
Comment (sidpremkumar) [reply src/services/Tracking/Tracking.service.ts:53]: Tried that originally but it was not sending events on calls to `cleanExit`. Happy to hop on a gmeet and try debugging together, but that was the code I had originally and was running into issues. 

I believe that the function call `this.visitor.event` is awaited, but not the callback possibly? I'm not sure tbh.

File: src/services/Tracking/Tracking.service.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+            await zliArgsCall;
+        }
+    }
+
+    /**
+     * Helper function to track a cli error.
+    */
+    public async TrackError() {
+        const zliErrorCall = new Promise<void>(async (resolve, _) => {
+            await this.visitor.event('zli-error', 'lt', (err: any) => {
```
[the remark this answers is not in the document]
Comment (sidpremkumar) [reply src/services/Tracking/Tracking.service.ts:75]: Yup thanks, this should have been `baseCommand`

File: src/services/Tracking/Tracking.service.ts (hunk 5 above)
Comment (sidpremkumar) [reply src/services/Tracking/Tracking.service.ts]: https://github.com/bastionzero/zli/pull/304/commits/d018c950cff0dacd67575e85a499fd5c710d8650

File: src/services/v1/token/token.messages.ts (hunk 4 above)
Comment (sidpremkumar) [reply src/services/v1/token/token.messages.ts]: https://github.com/bastionzero/zli/pull/304/commits/d018c950cff0dacd67575e85a499fd5c710d8650

File: src/services/v1/token/token.service.ts (hunk 3 above)
Comment (sidpremkumar) [reply src/services/v1/token/token.service.ts:2]: https://github.com/bastionzero/zli/pull/304/commits/d018c950cff0dacd67575e85a499fd5c710d8650

File: src/services/config/config.service.ts (hunk 1 above)
Comment (sidpremkumar) [reply src/services/config/config.service.ts]: https://github.com/bastionzero/zli/pull/304/commits/d018c950cff0dacd67575e85a499fd5c710d8650

File: src/http-services/token/token.http-services.ts [hunk 12]
```diff
@@ -13,7 +13,12 @@ export class TokenHttpService extends HttpService
         super(configService, 'api/v2/token/', logger, false);
     }
 
-    public getMixpanelToken(): Promise<MixpanelTokenResponse>
+    public getGAToken(): Promise<TrackingTokenResponse>
+    {
+        return this.Get('GA-token', {});
```
[the remark this answers is not in the document]
Comment (sidpremkumar) [reply src/http-services/token/token.http-services.ts]: https://github.com/bastionzero/zli/pull/304/commits/d018c950cff0dacd67575e85a499fd5c710d8650

File: src/services/Tracking/Tracking.service.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (tflpd) [reply src/services/Tracking/Tracking.service.ts]: Hm, I guess they do not really have support for TS, alright!
Feel free to resolve

File: src/services/Tracking/Tracking.service.ts (hunk 5 above)
Comment (tflpd) [reply src/services/Tracking/Tracking.service.ts]: I still see both the folder and the filename capitalized

Comment (tflpd) [reply src/services/Tracking/Tracking.service.ts]: My bad, it is just the folder actually

**Review by tflpd — DISMISSED** [review]

**Review by sidpremkumar — APPROVED** [review]

**Pull request merged**